### DISCUSSION TITLE
Is there a way of excluding a management group resource declaration?

### DISCUSSION THREAD

**Discussion by BenjaminEngeset** [Azure/AzOps#648] [Q&A] [answered]
Hello project contributors and maintainers. 

Is there currently a way of excluding a management group resource declaration? 

In a situation where the platform team does not have an `owner role` or `Microsoft.Resources/deployments/write` at management group.

`Reader` is given at management group and at subscription they are `contributors`. 

That works actually nice with AzOps; it is still able to traverse through management group since it can read it.

Problem is when it wants to push the tracked management resource declaration..

```
New-AzManagementGroupDeployment: The client 'xxx' with object id
'xxx' does not have authorization to perform
action 'Microsoft.Resources/deployments/write' over scope...
```
I see that `"Core.SkipResourceType"` is only for resource types which targets resource group scoped resources.

I understand that AzOps is intended and designed to be used with an owner role, but I had a slight hope that there might be some sort of way to get it to overlook this management resource declaration that got pulled in.

Thanks in advance =)

/b

**Comment by daltondhcp** [comment, marked as answer]
Before we dig into potential solutions for your use case, it would be interesting to understand why you grant reader permissions at MG level if they are only supposed to operate at subscription level? If you remove MG permissions, folders for each subscription will be created directly under the root folder, and subject to correct roleAssignments there shouldn't be any permission issues 😀

**Reply by BenjaminEngeset** [reply]
@daltondhcp 

Thank you for taking your time to assist. 

The reason why is because I am not able to make it work on only subscriptions permissions either (owner or contributor).

![image](https://user-images.githubusercontent.com/99641908/171210959-66ac49ae-c82e-48ba-b582-5523fb8e0096.png)

It is able to deploy everything what `settings.json` is referring to, expect subscription that happends around 5 minutes into the deployment.

![image](https://user-images.githubusercontent.com/99641908/171212472-8dd51655-e1c2-45d0-bf8b-51a9978a1730.png)
```

**Reply by BenjaminEngeset** [reply]
Under `deployment `

![image](https://user-images.githubusercontent.com/99641908/171213468-824f3bd6-f457-4416-b069-dbe51e934231.png)

`settings.json`

![image](https://user-images.githubusercontent.com/99641908/171213699-2117c827-4880-40d9-a76c-589a8df66ceb.png)

Tried with multiple different tenants, same issue occurs.

**Reply by daltondhcp** [reply]
This only happens after initial pull, right? Ideally (on GitHub it works like this), we don't deploy all resources back after initial pull, but that seems to be happening in your case with ADO. Since we just pulled the resources back from Azure it doesn't really require a push pack to the platform. 
If you try to update a single file, or for example create an RG post the initial pull/push operation, it works as expected?

Nevertheless, I think the solution here is to change the trigger for the Push pipeline to exclude the automated PRs coming from Pull. Obviously, it will still fail if you make manual changes to subscription.json, but then you will better understand why during validate :-)

**Reply by BenjaminEngeset** [reply]
@daltondhcp 

Yes. This happends after the first push (where you delete .github or .md to trigger git diff) , then pull is triggered, then push is triggered again where it wants to deploy everything from first pull.

What you suggested worked.

I canceled the pull job that started from being triggered from first pull. Added a random bicep deployment and it only deploys that.

It all makes sense now and I understand the behavior.

I agree with your proposition. Then you will have state in the branch, but it will not be deployed (that would fail due to lacking permissions).

Then when doing human interaction deployment from the AzOps repo, it will only deploy that since git diff against main (default) branch.

This also solves the when `subscription.json` is changed, it will not try to deploy it even when pull is tracking the new changes.

It will of course fail if someone from the platform team tries to manually edit it and merge it to main, but they should not do that anyways.

The only **real** trade off is not being able to deploy ` json` from pull (and `bicep` in the future), but that is rather more permission issue than AzOps design. Would have been nice to test the state that is in the code base, but..

Thank you for helping me with this headache, it was quite confusing even if it should have been quite obvious.

**Reply by daltondhcp** [reply]
Glad to hear that it helped. Why would you want to deploy back json or bicep directly coming from pull without first updating the autogenerated files  via a new branch/pr? 
Since it already comes directly from the platform so there would be nothing to deploy anyway? As mentioned, this is actually the default behavior on GitHub for the automated PRs.

**Reply by BenjaminEngeset** [reply]
Good point, it is identical so should have nothing to say.

Will also save deployment history that can cause confusion and noise for the team or other teams that might be involved in the organization. 

I am happy to see that there are many cases where AzOps can be used, not only high level permission.

If you only give the security principal access from RGs and down, will AzOps tackle that as well and understand the context?

**Reply by daltondhcp** [reply]
Yes, it works exactly the same as with subscription scoped SPNs, but RG folders will only be created subject to the RBAC permissions. (Similar to how it works when you have RG scoped permissions in a subscription and are using the Azure portal)

**Reply by BenjaminEngeset** [reply]
Do you have the example how it was solved for GH so I can try to see if same logic can be applied for ADO? Was trying to check the examples, but did not see it super clear (most probably due to lacking GH knowledge). Just trying to put together how to differentiate on this when both the human interaction and pull are both merging into main branch:

```
trigger:
  branches:
    include:
      - main
  paths:
    include:
      - root
```

**Reply by daltondhcp** [reply]
For GitHub it is actually the default behavior, so no extra config needed there. 
For ADO, you need to add a [condition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml), like commit message or source branch. Unfortunately the [`[skip ci]` method ](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/azure-repos-git?view=azure-devops&tabs=yaml#skipping-ci-for-individual-pushes)doesn't work on PR merge. 

See simple example how it can be implemented below: 

**push.yml**
```yaml
jobs:  
  - job: push
    condition: not(contains(variables['Build.SourceVersionMessage'], 'from automated into main'))
```
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/16622613/171275525-b4ba3000-d68a-4ad9-b5af-40c1b2cbbc7a.png">


Obviously, you can use other [build variables](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#build-variables-devops-services) or messages than the one above, but you get the picture 😄

**Reply by BenjaminEngeset** [reply]
@daltondhcp 

You are the man, thank you! I can't believe I was able to overlook this in the documentation.

I was trying to edit the default commit message to add [skip ci], but it never worked and I saw similar issues in the community.

Again, truly grateful.
